### PR TITLE
[DI] Don't track merged configs when the extension doesn't expose it

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24020
| License       | MIT
| Doc PR        | -

This is driving me crazy :)

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
So, this happens when the DI extension doesn't call `Extension::processConfiguration()`, which is the case in PropelBundle (fixed in propelorm/PropelBundle#461.)

Shouldn't be common, so I wouldn't recommend a new release (yet.)

**Comment by gharlan** [comment]
@nicolas-grekas Thanks! I can confirm:

1. It works with your PropelBundle fix, without this pr
2. It works with your pr without changing PropelBundle
3. And of course, it also works with both changes together.

**Review by chalasr — APPROVED** [review]
👍  Do not deserve an immediate release to me either

**Comment by jeremyFreeAgent** [comment]
Hi guys, thanks for the fix!

What about release that one since every bundle without the fix may have issue?

@gharlan confirms that this PR fix the issue without change in the bundles.

**Comment by nicolas-grekas** [comment]
@jeremyFreeAgent if you have the issue (do you?), you'd better fix it in your bundle also because it's current behavior will prevent correct tracking of ENV bars anyway.

**Comment by jeremyFreeAgent** [comment]
@nicolas-grekas I was not talking about _my_ bundle but the _community_ bundles. I mean okay we can PR every bundle we spot with that issue but what about _basic users of Symfony_ that will only update Symfony and expect everything still works?

BTW I have PR the fix in https://github.com/sensiolabs/SensioLabsConnectBundle/pull/17 and It was merged :)

**Comment by samvdb** [comment]
I can confirm the same bug using 8p/guzzle-bundle.

**Comment by nicolas-grekas** [comment]
@samvdb can you submit a PR on https://github.com/8p/GuzzleBundle/blob/master/DependencyInjection/GuzzleExtension.php to fix that extension the same way as done in https://github.com/sensiolabs/SensioLabsConnectBundle/pull/17?

**Comment by nicolas-grekas** [comment]
BTW, I'll propose to deprecate not calling Extension::processConfiguration in 3.4, and throw an exception in 4.0

**Comment by samvdb** [comment]
@nicolas-grekas PR is open, can confirm this fixes the issue.
https://github.com/8p/GuzzleBundle/pull/120

**Comment by stof** [comment]
@nicolas-grekas could we detect that this method was not called, and use the previous behavior for replacing env placeholders in this case ? This would not solve the shadowing of the config in case a variable is not used anymore in the end. But it would mean that the contain keeps working when the config is now shadowed.
Otherwise, it means we broke BC in a patch version.

**Comment by nicolas-grekas** [comment]
@stof see patch, that should be what L97 does.

File: src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php [hunk 1]
```diff
@@ -92,12 +92,16 @@ public function __construct(parent $parameterBag)
...
+        if (!$config = $extension->getProcessedConfigs()) {
+            // Extension::processConfiguration() wasn't called, we cannot know how configs were merged
+            return;
+        }
         $this->processedEnvPlaceholders = array();
 
-        // serialize config to catch env vars nested in object graphs
-        $config = serialize($extension->getProcessedConfigs());
+        // serialize config and container to catch env vars nested in object graphs
+        $config = serialize($config).serialize($container->getDefinitions()).serialize($container->getAliases()).serialize($container->getParameterBag()->all());
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php]: aliases don't need to be serialized. They only have a target and a public flag, and these cannot use parameters.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php]: They can somehow, see https://github.com/nicolas-grekas/symfony/blob/a8e6aac2f58ca3f12c931f20a6de29ef46686dc8/src/Symfony/Component/DependencyInjection/Compiler/ResolveParameterPlaceHoldersPass.php

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**